doc/make_roa.1
MAKE_ROA (1)             ROA-PKI                                   MAKE_ROA(1)

NAME

        make_roa - makes a ROA

SYNOPSIS

        make_roa  -r roadatafile -c certfile -k keyfile -o outfile [-v vfile]

        make_roa parameterfile

DESCRIPTION

          This program (in the roa-utils directory) creates a ROA from these
          parameters supplied:

            -r roadatafile where roadatafile is he name of a file containing
               data for the roa in text form, including AS#, IPv4 prefixes
               and IPv6 prefixes in the format created by the read_roa
               program (q.v).

            -c certfile where certfile is the name of the file containing
               the certificate of the entity that will sign the ROA

            -k keyfile where keyfile is the name of a file containing the
               private key to be used for signing the ROA.

            -o outfile where outfile is the name of the file to which the ROA
               is to be written

            -v vfile where human-readable version of outfile is to be written.

        The parameters may be in any order.

            As an alternative, the program may be run with a single parameter
        naming a file which contains the above-named parameters separated by
        white space, e.g. the file might contain:

                -c mycertificate -k mykey -o some.roa -r roa.txt -v some.raw

            When the program completes successfully, it ends with the message

                 Created outfile

        If there is any error, the program terminates with one of several
        messages:

          Missing xxx
          invalid parameter yyy
          Couldn't open filename
          Error reading filename
          Error signing with filename
          Error writing wilename

        where xxx is the name of a parameter, yyy is a parameter given, and
        filename is the name of a file in the parameter list.


ChangeLog
0.12, not yet released

	* Miscellaneous code clean-up and minor bug fixes.
	* Adjust README to instruct users to build a shared version of
	  OpenSSL.
	* Bump copyright year to 2016.
	* Fix unsafe use of snprintf().
	* A compiler with C99 support is now required.
	* Build system improvements.  Autoconf 2.60 is now required.
	* Test suite improvements.

0.11, released 2015-06-12

	* Fix a bug where ROAs with hundreds of ROAIPAddresses or more would be
	  truncated.
	* Add support for new-style Trust Anchor Locators (TALs)
	  [draft-ietf-sidr-rfc6490-bis-03].
	* Fix a potential buffer overflow in code that handles untrusted input.
	* Fix performance scaling of rpki-rtr-daemon. The rpki-rtr-daemon
	  previously had poor performance with very large rtr_incremental or
	  rtr_full tables.
	* Decrease the likelihood of a malformed rsync URI resulting in a
	  successful attack, by passing the --protect-args option to rsync.
	* Improve conformance to [RFC6810] by sending a No Data Available PDU
	  instead of a Cache Reset PDU in response to a serial query for which
	  we don't have data yet.
	* Fix expired certificates in the test suite.
	* Fix a potential infinite loop in one of the self tests.
	* Fix a compatibility bug in the test suite, where the tests were
	  failing on several popular systems that use a particular version of
	  netcat.

0.10, released 2014-02-25

	* Add beta support for running rpstir in 64-bit mode.
	* Add beta support for running rpstir on Mac OS X.
	* Significantly improve performance of incremental updates by adding an
	  rsync flag to preserve file modification times.
	* Fix multiple potential SQL-injection bugs.
	* Support newer versions of rsync and automake.
	* Remove support for RPSL due to lack of demand.
	* Update conformance cases (see doc/conformance-cases).
	* The ./configure script no longer prints a misleading error message
	  when it's run outside of a git directory.
	* Add tests with AS numbers that don't fit into 16 bits.
	* Reject trust anchor CA certificates that have resources marked
	  inherit, as specified in RFC 6490, Section 2.2.
	* Verify that rpstir compiles with clang, and fix all of clang's
	  warnings.
	* In the statistics collection mode, add support for incremental updates
	  and for running multiple simultaneous statistics collections.
	* Change handling of CRLs that list syntactically invalid serial
	  numbers. These CRLs are now accepted, but the certificates with
	  syntactically invalid serial numbers are still not accepted.

0.9, released 2013-10-31

	* Add many more conformance test cases. (See doc/conformance-cases for
	  the full list.) These test cases can be used by relying party
	  software to test compliance with published RFCs and Internet-Drafts.
	  Additionally, relying party software that passes the tests can be
	  used to test the output of a Certification Authority.
	* Fix bugs found by the above test cases.
	* Add a pseudo-random factor to the calculation of how long to wait
	  before retrying an rsync connection. This should help prevent many
	  relying parties from hitting the same server at the same time.
	* Implement basic support for collecting statistics of the RPKI over
	  time.

0.8, released 2013-07-25

	* Support Ghostbusters Records [RFC 6493].
	* Fix a SQL injection bug when a filename in a manifest contains certain
	  characters.
	* Fix a bug where CRLs were not being marked as invalid when their
	  issuers became invalid and manifests were not being marked as invalid
	  when their associated EE certificates became invalid.
	* Fix the query client (rpstir-query) to not consider it an error
	  when no results are returned for a query.
	* Require ROA EE certificates to explicitly specify IP resources.
	  That is, disallow RFC 3779 "inherit" bit in ROA EE certificates.
	* Add functionality to rpstir-results for outputting simple lists
	  of good and bad URIs, to simplify comparison testing.
	* On installation, create ${bindir} if it does not already exist.
	* Improve README based on user feedback.

0.7, released 2013-04-10

	* Add preliminary support for adding files from the local cache. In the
	  future, this could be used to quickly deploy additional relying party
	  machines without each one needing to download all files from the
	  global RPKI.
	* Fix a buffer overflow when manifests list filenames longer than 199
	  bytes.
	* Fix a read of uninitialized memory when the first file listed on a
	  manifest has an unknown extension.
	* Don't allow empty sets of IP resources in certificates.
	  (http://tools.ietf.org/html/rfc6487#section-4.8.10)
	* Don't allow Subsequent Address Family Identifier (SAFI) values in IP
	  resources. (http://tools.ietf.org/html/rfc6487#section-4.8.10)
	* Fix bugs in the ROA validation code where AS numbers equal to zero
	  or larger than 2^31 - 1 were rejected.

0.6, released 2013-02-15

	* Simplify the commands used during normal operation. The commands
	  are now named after what the user intends to do, instead of after the
	  internal component we used to implement them. Some of the scripts are
	  reorganized so that each script does one function (e.g., synchronize
	  with remote repositories) instead of presenting an interface to one
	  internal component (e.g., the component that validates objects fed to
	  it by other components while syncronizing with remote repositories).
	* Use a single configuration file syntax instead of a mix of
	  environment variables and multiple incompatible configuration files.
	  Additionally, provide an example configuration file with comments
	  describing each option. This should make rpstir easier to set up.
	* Switch from one Makefile per directory to a single Makefile. Aside
	  from speeding up parallel compilation, this lets us simplify
	  internal dependencies. With one Makefile per directory, we had to
	  carefully order the directories such that nothing in one directory

roa-utils/make_roa.c
/*
 * $Id: make_roa.c 453 20078-07-25 15:30:40Z gardiner $ 
 */


#include "roa_utils.h"
#include "manifest.h"
#include "cryptlib.h"
#include <stdio.h>
#include <sys/types.h>
#include <time.h>

char *msgs[] = {
    "Created %s\n",
    "Can't open %s\n",          // 1
    "Invalid prefix %s\n",
    "Missing %s\n",             // 3
    "%s covers AS number and v4 and v6 files\n",
    "Error in %s when signng\n",        // 5
    "Error writing %s\n",
    "Invalid parameter %s\n",   // 7
};

static int fatal(
    int msg,
    char *paramp)
{
    fprintf(stderr, msgs[msg], paramp);
    exit(msg);
}

static int prefix2roa(
    struct ROAIPAddress *roaIPAddrp,
    char *prefixp,
    int family)
{
    char *c;
    int pad = 0,
        siz;
    for (c = prefixp, siz = family; *c >= ' ' && *c != '/'; c++)
    {
        if (family == 1)
        {
            if (*c == '.')
                siz += 1;
        }
        else if (*c == ':')
        {
            if (c[1] != ':')
                siz += 2;
            else if (pad)
                fatal(2, prefixp);
            else
                pad = 1;
        }
    }
    if (pad)
        pad = 16 - siz;
    uchar *buf = (uchar *) calloc(1, siz + pad + 2);
    uchar *b;
    int i;
    for (c = prefixp, b = &buf[1]; *c >= ' ' && *c != '/'; c++)
    {
        if (family == 1)
        {
            sscanf(c, "%d", &i);
            *b++ = i;
        }
        else if (*c == ':')
        {
            int j;
            for (j = 0; j < pad; *b++ = 0, j++);
        }
        else
        {
            sscanf(c, "%x", &i);
            *b++ = (uchar) (i >> 8);
            *b++ = (uchar) (i & 0xFF);
        }
        while (*c > ' ' && *c != '.' && *c != ':' && *c != '/')
            c++;
        if (*c == '/')
            break;
    }
    if (*c == '/')
    {
        c++;
        sscanf(c, "%d", &i);
        while (*c >= '0' && *c <= '9')
            c++;
        siz += pad;
    }
    int lim = (i + 7) / 8;
    if (siz < lim)
        fatal(2, prefixp);
    else if (siz > lim)
    {
        b--;
        if (*b)
            fatal(2, prefixp);
        siz--;
    }
    i = (8 * siz) - i;          // i = number of bits that don't count
    uchar x,
        y;
    for (x = 1, y = 0; x && y < i; x <<= 1, y++)
    {
        if (b[-1] & x)
            fatal(2, prefixp);
    }
    buf[0] = i;
    write_casn(&roaIPAddrp->address, buf, siz + 1);
    if (*c == '^')
    {
        sscanf(++c, "%d", &i);
        write_casn_num(&roaIPAddrp->maxLength, (long)i);
    }
    return siz;
}

static void do_family(
    struct ROAIPAddrBlocks *roaBlockp,
    int famnum,
    int x,
    FILE * str)
{
    struct ROAIPAddressFamily *roafamp =
        (struct ROAIPAddressFamily *)inject_casn(&roaBlockp->self, x);
    uchar family[2];
    char *c,
        nbuf[256];
    family[0] = 0;
    family[1] = famnum;
    write_casn(&roafamp->addressFamily, family, 2);

    int numaddr;
    for (numaddr = 0; fgets(nbuf, sizeof(nbuf), str); numaddr++)
    {
        if (nbuf[0] > ' ')
            return;
        for (c = nbuf; *c == ' '; c++);
        struct ROAIPAddress *roaIPaddrp =
            (struct ROAIPAddress *)inject_casn(&roafamp->addresses.self,
                                               numaddr);
        prefix2roa(roaIPaddrp, c, famnum);
    }
}

int main(
    int argc,
    char **argv)
{
    struct ROA roa;

    if (argc < 2)
    {
        fprintf(stderr, "Need EITHER a parameter file name\n");
        fprintf(stderr, "OR -c certificateFile -k keyfile -o outputfile\n");
        fprintf(stderr, "    -r read_roa_outputfile or\n");
        return 0;
    }
    char *b,
       *c,
       *e,
       *buf = (char *)0;
    char *certfile,
       *keyfile,
       *outfile,
       *readroafile,
       *vfile;
    char **pp,
      **tp;
    certfile = keyfile = outfile = readroafile = vfile = (char *)0;
    if (argc == 2 && argv[1][0] != '-') // parameter file
    {
        int fd = open(argv[1], O_RDONLY);
        if (fd < 0)
            fatal(1, argv[1]);
        int lth = lseek(fd, 0, SEEK_END);
        lseek(fd, 0, SEEK_SET);
        buf = (char *)calloc(1, lth + 1);
        read(fd, buf, lth);
        int numargs;
        for (e = buf; *e; e++)
            if (*e <= ' ')
                *e = 0;
        for (b = buf; b < e && *b == 0; b++);   // go to first parameter
        for (numargs = 0; b < e; numargs++)
        {
            while (b < e && *b != 0)
                b++;
            while (b < e && *b == 0)
                b++;
        }
        pp = (char **)calloc(1, (numargs + 1) * sizeof(char *));
        for (b = buf; b < e && *b == 0; b++);   // go to first parameter
        for (tp = pp; numargs-- > 0; tp++)
        {
            while (b < e && *b == 0)
                b++;
            *tp = b;
            while (b < e && *b > 0)
                b++;
        }
    }
    else
        pp = &argv[1];          // parameters as argvs
    for (; *pp; pp++)
    {
        c = *pp;
        if (*c != '-')
            fatal(7, c);
        b = *(++pp);
        if (*(++c) == 'c')
            certfile = b;
        else if (*c == 'k')
            keyfile = b;
        else if (*c == 'o')
            outfile = b;
        else if (*c == 'r')
            readroafile = b;
        else if (*c == 'v')
            vfile = b;
        else
            fatal(7, &c[-1]);
    }
    if (!certfile)
        fatal(3, "certificate file");
    if (!outfile)
        fatal(3, "output file");
    if (!keyfile)
        fatal(3, "key file");
    if (!readroafile)
        fatal(4, "readroafile");
    FILE *str = fopen(readroafile, "r");
    if (!str)
        fatal(1, readroafile);
    char nbuf[256];
    long asnum = -1;
    while (fgets(nbuf, 128, str))
    {
        if (!strncmp(nbuf, "AS#", 2))
        {
            for (c = &nbuf[3]; *c == ' '; c++);
            sscanf(c, "%ld", &asnum);
            break;
        }
    }
    fseek(str, 0, SEEK_SET);
    ROA(&roa, (ushort) 0);
    write_objid(&roa.contentType, id_signedData);
    struct SignedData *sgdp = &roa.content.signedData;
    write_casn_num((struct casn *)&sgdp->version, 3);
    struct AlgorithmIdentifier *algidp =
        (struct AlgorithmIdentifier *)inject_casn(&sgdp->digestAlgorithms.self,
                                                  0);
    write_objid(&algidp->algorithm, id_sha256);
    write_casn(&algidp->parameters.sha256, (uchar *) "", 0);
    write_objid(&sgdp->encapContentInfo.eContentType,
                id_routeOriginAttestation);
    struct RouteOriginAttestation *roap = &sgdp->encapContentInfo.eContent.roa;
    write_casn_num(&roap->asID, asnum);
    int x = 0;
    while ((c = fgets(nbuf, sizeof(nbuf), str)) && strncmp(nbuf, "IPv4", 4));
    if (c)
        do_family(&roap->ipAddrBlocks, 1, x++, str);
    fseek(str, 0, SEEK_SET);
    while ((c = fgets(nbuf, sizeof(nbuf), str)) && strncmp(nbuf, "IPv6", 4));
    if (c)
        do_family(&roap->ipAddrBlocks, 2, x, str);
    struct Certificate *certp =
        (struct Certificate *)inject_casn(&sgdp->certificates.self, 0);
    if (get_casn_file(&certp->self, certfile, 0) < 0)
        fatal(2, certfile);
    if ((c = signCMS(&roa, keyfile, 0)))
        fatal(5, c);
    if (put_casn_file(&roa.self, outfile, 0) < 0)
        fatal(6, outfile);
    if (vfile)
    {
        int lth = dump_size(&roa.self);
        char *dbuf = (char *)calloc(1, lth + 8);
        dump_casn(&roa.self, dbuf);
        FILE *vstr = fopen(vfile, "w");
        fputs(dbuf, vstr);
        fclose(vstr);
        free(dbuf);
    }
    if (buf)
        free(buf);
    fatal(0, outfile);
    return 0;
}
